managed/models/user_roles_model.go

```diff
@@ -34,6 +34,16 @@ type UserRoles struct {
 	UpdatedAt time.Time `reform:"updated_at"`
 }
 
+// SetEntityID sets user ID
```

Comment:
Could we refactor this model to include entity ID, entity type, and role here instead of creating another model which will be pretty similar to this one?

Reply:
in that case we wouldn't need all these hacks.

Reply:
Usually I try to avoid polymorphic relationships in database schemas unless there's great support in the ORM for them.
Without a great support, it's easy to overlook issues with queries for developers without experience in this part of codebase.

Personally I'd prefer individual tables for the relationships. Let me know if we shall move forward with `entity_type`, `entity_id` and `role_id` table instead. Happy to refactor.

Reply:
even if polymorphic relationships don't exist in a database schema, they still exist in the code. it makes the code much harder to read. We use these kinds of columns everywhere, agent_type, service_type and etc.
maybe we can get another opinion from other developers?
@artemgavrilov or @PavelKhripkov your thoughts?

Reply:
I can't understand what you're discussing without the context? What is the entity, and why method `SetEntityID` uses `userID` and an argument?

Reply:
We need to store the following relationships:
- `user_id` to `role_id` (one to many)
- `team_id` to `role_id` (one to many)

We can either have polymorphic relationships in a single table, such as with `entity_id`, `entity_type` and `role_id` colums where `entity_id` refers to the appropriate entity based on `entity_type`.

Or we can have two separate tables `user_roles` and `team_roles` and handle the rest in the code similar to how the PR handles it.

Reply:
I'm for polymorphic relationships because here are too many hacks around the usage of multiple tables.

Reply:
@michal-kralik as I understood, you're only preparing to introduce Team entity. And it's not clear what is the relationship between User and Team models?

Reply:
A user can belong to one or many teams.
Each user and team can have a set of roles assigned and the resulting set of roles for a user is a union of all the roles from a user and a team.

Reply:
Polymorphic relationships usually used to store entities of the same level. E.g. you have a table with history of actions and then you can store records like:
`id=1, type=service, action=delete, time=123`
`id=8, type=node, action=add, time=456`

There are no relations between records in the `history` table. But as I see you're trying to store in a single table entities of different levels, of different meaning and with relations between them. To me it looks extremly confusing. By the meaning entity `Team` is close to `Role` rather than to `User` because it provides access level to the User. But storing `Role` and `Team` in a single table not much better.

Reply:
a role can be assigned to a user as well as it can be assigned to a team.

Reply:
Role can be assigned to a user as well as Team can be assigned to a User)

Reply:
yes, but a team and a role aren't the same :)
a team is a group of users whereas a role is a group of permissions :)

Reply:
Right. And if we're thinking about Team as about marker or label then there is no need to add new entity, we can just add labels to User or Role. But I imagine Team as not something to only group Roles. Team could have other properties hepling to manage groups of Users. And if we put User and Team in the same table we need much efforts later to extend Teams functionality.

Reply:
currently, Teams and users are stored in the Grafana database and this table is used only to assign a role to a team or a user. It's a table for many-to-many connections. There shouldn't be any new columns in the future.

Reply:
Ah, I see, you already have a data model and trying to implement new functionality on top of that. Then it's not so bad as I thought)

Reply:
Yes)